### PR TITLE
♻️ HTTP Rewrite of networking library

### LINKED ISSUE #28: :key: Logout route
# 👮‍♂️ What's needed
Create a logout route to disconnect  the client from the server

# 📜 Sources
no sources

### LINKED ISSUE #29: 💾  Persistent data
# 👮‍♂️ What's needed
Save and load data between launches

## 📋 Subtasks
- [x] Save data
- [x] Load data

# 📜 Sources
No sources

### LINKED ISSUE #30: :key: Login route
# 👮‍♂️ What's needed
Create the login route, to make the connection of the client to the server

# 📜 Sources
no sources

### PR DESCRIPTION
**Pull request by axel-eck**
I rewrote the networking library so it's now compliant with http 1.0

I also updated code that depended on this library.

This pull request also implements #30 #28 #29

### CODE REVIEW

File: client/Makefile [hunk 1]
```diff
@@ -23,7 +23,8 @@ NAME 			=   myteams_cli
 CC				= 	gcc
 FLAGS			= 	-W -Wall -Wextra --coverage -I./include \
 					-L../libs/myteams/ -L../libs/  -lnetwork -ljson \
-					-I../include -I../libs/myteams -lmyteams
+					-I../include -I../libs/myteams \
+					$(shell [ `uname` == "Linux" ] && echo "-lmyteams")
```
Comment (Marius-P1) [inline client/Makefile:27]: The lib myteams still don't work on your mac ?

File: client/src/client.c [hunk 2]
```diff
@@ -97,29 +98,21 @@ int main_loop(client_t *client)
...
-        return 84;
-    }
     if (is_valid_args(ac, av) == 84) {
         print_help();
         return 84;
     }
-    host->ip = strdup(av[1]);
-    host->port = atoi(av[2]);
-    api_handler = network_create_api_handler(host);
+    api_handler = api_client_init(av[1], atoi(av[2]));
```
Comment (Marius-P1) [inline client/src/client.c:115]: Does it check if the server exist or not ?

File: common/uuid/src/uuid.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "uuid/uuid.h"
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+
+bool uuid_compare(const uuid_t uu1, const uuid_t uu2)
+{
+    for (int i = 0; i < 16; i++)
```
Comment (Marius-P1) [inline common/uuid/src/uuid.c]: I think you can use `size_t` here but it's not mandatory

File: server/Makefile [hunk 4]
```diff
@@ -40,8 +66,9 @@ all:		$(NAME)
 $(NAME):	$(OBJS)
 # Link the object files
 		@printf "$(RUNNING) $(BLUE) 🔗  Linking$(RESET)"
-		@$(CC) -o $(NAME) $(OBJS) $(FLAGS) \
-		&& printf "\r$(SUCCESS)\n" || printf "\r$(FAILURE)\n"
+		@$(CC) -o $(NAME) $(OBJS) $(FLAGS) >> $(LOG) 2>&1 \
+		&& printf "\r$(SUCCESS)\n" || (printf "\r$(FAILURE)\n" && cat $(LOG) \
+		&& exit 1)
```
Comment (Marius-P1) [inline server/Makefile:71]: Why exit, don't you have a if after to check if the process was successful ?

File: server/src/routes/login.c [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    roundtable_client_t *client = NULL;
+    json_object_t *response_body = json_object_create(NULL);
+    json_string_t *uuid_json = NULL;
+    char *response_body_str = NULL;
+
+    if (body == NULL || !json_object_has_key(body, "username"))
+        return create_error(400, "Bad Request", "JSON", "Missing ursername");
+    username = ((json_string_t *) json_object_get(body, "username"))->value;
+    client = get_or_create_client(server, username, &response);
+    uuid_json = json_string_create("uuid", uuid_to_string(client->uuid));
```
Comment (Marius-P1) [inline server/src/routes/login.c]: Please use the correct key as written in the RFC :eyes:

File: server/src/main.c [hunk 6]
```diff
@@ -6,9 +6,53 @@
...
+    if (access("./data.json", F_OK) != -1)
+        load_data(server, "./data.json");
+    bind_routes(server);
+    get_server(true, server);
+    sigaction(SIGINT, &(struct sigaction){.sa_handler = &stop}, NULL);
+    router_start(server->router);
+    save_data(server, "./data.json");
+    destroy_server(server);
     return 0;
 }
```
Comment (Marius-P1) [inline server/src/main.c:58]: I recommend using a different file (like a server.c) to put the initial logic and only call one function in the main for unit test purpose. But it's not a mandatory

File: server/src/routes/logout.c [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+#include "server_utils.h"
+
+response_t logout_route(request_t *request, void *data)
+{
+    roundtable_server_t *server = (roundtable_server_t *)data;
+    json_object_t *json = (json_object_t *) json_parse(request->body);
+    char *user_uuid = NULL;
+    roundtable_client_t *client = NULL;
+
+    if (strcmp(request->route.method, "POST") != 0)
```
Comment (Marius-P1) [inline server/src/routes/logout.c:20]: You didn't check the method in login.c. It would be better to do it

File: server/src/routes/messages/index.c
Comment (Marius-P1) [inline server/src/routes/messages/index.c:1]: I'm not a fan of all the empty files

File: tests/uuid/test_uuid.c [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    cr_assert_eq(uuid_compare(uu1, uu2), true);
+}
+
+Test(uuid, compare_same)
+{
+    uuid_t uu1 = {0};
+    uuid_t uu2 = {0};
+
+    for (int i = 0; i < 16; i++) {
```
Comment (Marius-P1) [inline tests/uuid/test_uuid.c:25]: It would be better to use `size_t` but it's not a mandatory

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
Somes changes to do

File: client/Makefile (hunk 1 above)
Comment (axel-eck) [reply client/Makefile:27]: No sadly

File: server/Makefile (hunk 4 above)
Comment (axel-eck) [reply server/Makefile:71]: It's only launched when the commands failed due to "||"

File: server/src/routes/logout.c (hunk 7 above)
Comment (axel-eck) [reply server/src/routes/logout.c:20]: I don't have enough line but I will try

File: server/src/routes/messages/index.c
Comment (axel-eck) [reply server/src/routes/messages/index.c:1]: Me neither

File: server/src/main.c (hunk 6 above)
Comment (axel-eck) [reply server/src/main.c:58]: We will let it like that for the moment but I keep this in mind for future update

**Review by Marius-P1 — DISMISSED** [review]
👍

**Review by 6im0n — DISMISSED** [review]
Nice ! good for me :100:

**Review by 6im0n — APPROVED** [review]

**Pull request merged**